crates/syn2mas/src/synapse_reader.rs

```diff
@@ -0,0 +1,281 @@
+        }
+
+        Ok(Self {
+            conn: synapse_connection,
+        })
+    }
+
+    /// Finishes the Synapse reader, committing the transaction.
+    ///
+    /// # Errors
+    ///
+    /// Errors are returned under the following circumstances:
+    ///
+    /// - An underlying database error whilst committing the transaction.
+    pub async fn finish(self) -> Result<(), Error> {
+        // TODO enforce that this is called somehow.
+
+        query("COMMIT;")
+            .execute(self.conn)
+            .await
+            .into_database("end transaction")?;
+        Ok(())
+    }
+
+    /// Counts the rows in the Synapse database to get an estimate of how large the migration is going to be.
+    ///
+    /// # Errors
+    ///
+    /// Errors are returned under the following circumstances:
+    ///
+    /// - An underlying database error
+    pub async fn count_rows(&mut self) -> Result<SynapseRowCounts, Error> {
+        // TODO no need for query builder here
+        let (sql, args) = Query::select()
+            .expr(Expr::val(1).count())
+            .and_where(Expr::col(ExtraSynapseUserIden::AppserviceId).is_null())
+            // TODO support migrating at least skeleton records for guests
+            .and_where(Expr::col(ExtraSynapseUserIden::IsGuest).eq(0))
+            .from(SynapseUserIden::Table)
+            .build_sqlx(PostgresQueryBuilder);
+        let users = query_with(&sql, args)
+            .fetch_one(&mut *self.conn)
+            .await
+            .into_database("counting Synapse users")?
+            .try_get::<i64, _>(0)
+            .into_database("couldn't decode count of Synapse users table")?;
+
+        Ok(SynapseRowCounts { users })
+    }
+
+    /// Reads Synapse users, excluding application service users (which do not need to be migrated), from the database.
+    pub fn read_users<'a, 'ret>(
+        &'a mut self,
+    ) -> impl Stream<Item = Result<SynapseUser, Error>> + 'ret
+    where
+        'conn: 'a,
+        'a: 'ret,
+    {
+        // TODO no need for query builder here
+        let (sql, args) = Query::select()
+            .columns([
+                SynapseUserIden::Name,
+                SynapseUserIden::PasswordHash,
+                SynapseUserIden::Admin,
+                SynapseUserIden::Deactivated,
+                SynapseUserIden::CreationTs,
+            ])
+            .and_where(Expr::col(ExtraSynapseUserIden::AppserviceId).is_null())
+            // TODO support migrating at least skeleton records for guests
+            .and_where(Expr::col(ExtraSynapseUserIden::IsGuest).eq(0))
+            .from(SynapseUserIden::Table)
+            .build_sqlx(PostgresQueryBuilder);
+
+        let conn = &mut *self.conn;
+
+        // The async stream macro works around an issue where the QueryAs output stream borrows the SQL.
+        // See: https://github.com/launchbadge/sqlx/issues/1594#issuecomment-1493146479
```

Comment:
That would be fine if the query was `'static`, right?

Reply:
afaik yes — do you think it's better to hardcode the query (even though not using the `query!` macro) and so it would be `&'static` then?

Reply:
we hardcode queries in Synapse all the time, right? 😬 
If you'd like to keep the query builder, you could always stick the built query in a static `LazyLock`? But maybe that's a little verbose… As you want, your call

Reply:
on balance, happy enough to hardcode the query, since there will only be a few Synapse queries here and they won't really be getting changed once syn2mas is done, so the risk is a bit lower than usual.